### ISSUE TITLE
Unhelpful warnings produced when using source directive and subprocess measurement

### ISSUE THREAD

**Issue by nedbat** [coveragepy/coveragepy#96]
*Originally reported by* **Geoff Bache (Bitbucket: [geoffbache](https://bitbucket.org/geoffbache), GitHub: Unknown)**

----------------------------------------

When I specify my source in the [run] section of the rc file, and run coverage by setting COVERAGE_PROCESS_START, I get a lot of spurious warnings printed on standard error.

It says "Source module %s was never encountered." and occasionally "No data was collected."

Obviously not every subprocess will necessarily load every source module, and indeed it's perfectly OK to have some python subprocess somewhere which doesn't run any of my code.

The first warning is also a pain when running multiple tests in different processes: I want to have one rc file saying where my source is without needing every test to load every module.

I'd question whether these warnings are useful at all: but at least there should be some way to disable them, and preferably they should be off by default.

----------------------------------------
- Bitbucket: https://bitbucket.org/ned/coveragepy/issue/96

**Comment by nedbat** [comment]
I definitely see your points here.  The purpose of the warnings was to help people new to coverage to diagnose why they weren't getting the results they expected.  So I'm going to leave them enabled by default.  We can either disable them entirely with an rcfile setting, or provide a way to disable them individually, which is probably overkill.

**Comment by nedbat** [comment]
*Original comment by* **Geoff Bache (Bitbucket: [geoffbache](https://bitbucket.org/geoffbache), GitHub: Unknown)**

----------------------------------------

OK. Couldn't you achieve the first point with some kind of --debug / --verbose option? (i.e. documented as "enable this if you don't understand what's going on"?) 

I see your point re newbies but I think it isn't good to have things enabled by default that will often just be wrong. As far as I can see, they're only really useful if you're testing a single process in their current form.

**Comment by nedbat** [comment]
*Original comment by* **wigbam (Bitbucket: [wigbam](https://bitbucket.org/wigbam), GitHub: [wigbam](https://github.com/wigbam))**

----------------------------------------

This issue caused some pain to us as well. We are using pytest (and therefore pytest-cov and cov-core) to run tests within our company. When pytest-cov runs it creates coverage controller with explicit set of modules to collect coverage for, which is then also applied to all sub-processes (as far as my understanding goes).

Since a few of our tests parse the output from a sub-process they were failing due to warnings from coverage being present in the output. In the end I had to patch coverage to provide a possibility to disable all coverage warnings by setting an environment variable.

Therefore, in ideal case I would vouch for disabling coverage warnings for all sub-processes by default and only display them when aggregated output is generated at the test session end. Not sure if this needs to go into pytest-cov/cov-core or into coverage itself, though.

**Comment by nedbat** [comment]
*Original comment by* **Daniel Blanchard (Bitbucket: [daniel_blanchard](https://bitbucket.org/daniel_blanchard), GitHub: Unknown)**

----------------------------------------

Is there a way to disable them in the rcfile?

**Comment by nedbat** [comment]
Issue #355 is asking for a way to suppress warnings, which I think is a good idea.

**Comment by nedbat** [comment]
*Original comment by* **Chris Withers (Bitbucket: [chrisw](https://bitbucket.org/chrisw), GitHub: [chrisw](https://github.com/chrisw))**

----------------------------------------

Is there yet a way to suppress these warnings:

Coverage.py warning: No data was collected.
Coverage.py warning: {modname} archivist was never imported.

I did have a look at #355, but seems pretty empty...

**Comment by nedbat** [comment]
*Original comment by* **Simon Ye (Bitbucket: [yesimon](https://bitbucket.org/yesimon), GitHub: [yesimon](https://github.com/yesimon))**

----------------------------------------

It seems like an easy-ish thing to add to the config to control `self._warn_no_data = True, self._warn_unimported_source = True` in `control.py`. 

I'm also getting these warnings run through pytest-cov, so a config variable should do the trick nicely.

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

For the record here is a proposed fix https://bitbucket.org/ned/coveragepy/pull-requests/110/add-no-warnings-to-run-cmd-to-silence/diff

**Comment by nedbat** [comment]
I liked the idea of individually suppressable warnings, rather than a huge hammer that turns them all off.  Other opinions?

**Comment by nedbat** [comment]
*Original comment by* **Loic Dachary (Bitbucket: [dachary](https://bitbucket.org/dachary), GitHub: [dachary](https://github.com/dachary))**

----------------------------------------

:-D what about **--no-warnings=no_data,unimported_source** to only set  


```
#!python

self._warn_no_data = False
self._warn_unimported_source = False
```

**Comment by nedbat** [comment]
Implemented in [70ca90597acd (bb)](https://bitbucket.org/ned/coveragepy/commits/70ca90597acd)

**Comment by nedbat** [comment]
This is in 4.4b1.
